lib/bolt/project_migrator/modules.rb

```diff
@@ -0,0 +1,212 @@
+# frozen_string_literal: true
+
+require 'bolt/project_migrator/base'
+
+module Bolt
+  class ProjectMigrator
+    class Modules < Base
+      def migrate(project, modulepath)
+        return true unless ENV['BOLT_MODULE_FEATURE'] && project.modules.nil?
+
+        @outputter.print_message "Migrating project modules\n\n"
+
+        config     = project.project_file
+        puppetfile = project.puppetfile
+        moduledir  = project.managed_moduledir
+
+        # Notify user to manually migrate modules if using non-default modulepath
+        if modulepath != project.modulepath
+          @outputter.print_migrate_step(
+            "Project has a non-default configured modulepath, unable to automatically "\
+            "migrate project modules. To migrate project modules manually, see "\
+            "http://pup.pt/bolt-modules"
+          )
+          true
+        # Migrate modules from Puppetfile
+        elsif File.exist?(puppetfile)
+          migrate_modules_from_puppetfile(config, puppetfile, moduledir, modulepath)
+        # Migrate modules to updated modulepath
+        else
+          consolidate_modules(modulepath)
+          update_project_config([], config)
+        end
+      end
+
+      # Migrates modules by reading a Puppetfile and prompting the user for
+      # which ones are direct dependencies for the project. Once the user has
+      # selected the direct dependencies, this will resolve the modules, write a
+      # new Puppetfile, install the modules, and then move any remaining modules
+      # to the new moduledir.
+      #
+      private def migrate_modules_from_puppetfile(config, puppetfile_path, moduledir, modulepath)
+        require 'bolt/puppetfile'
+        require 'bolt/puppetfile/installer'
+
+        # Migrating modules includes prompting. Do nothing if stdin is not a tty.
+        unless $stdin.tty?
+          @outputter.print_migrate_step(
+            "stdin is not a tty, unable to automatically migrate project modules."\
+            "To migrate project modules manually, see http://pup.pt/bolt-modules"
+          )
+          return false
+        end
+
+        begin
+          @outputter.print_migrate_step("Parsing Puppetfile at #{puppetfile_path}")
+          puppetfile = Bolt::Puppetfile.parse(puppetfile_path)
+        rescue Bolt::Error
+          @outputter.print_migrate_step("Unable to parse Puppetfile and automatically migrate modules.")
+          return false
+        end
+
+        # Prompt for direct dependencies
+        modules = select_modules(puppetfile.modules)
+
+        # Create new Puppetfile object
+        puppetfile = Bolt::Puppetfile.new(modules)
+
+        # Attempt to resolve dependencies
+        begin
+          @outputter.print_message('')
+          @outputter.print_migrate_step("Resolving module dependencies, this may take a moment")
+          puppetfile.resolve
```

Comment:
Hm, this seems to stack trace if we can't resolve the Puppetfile - for example, I tried migrating the Bolt puppetfile which errored on the local modules (canary, aggregate, etc.):
```
lucy ~/githubs/bolt/project ((e445890...)) $ be bolt project migrate
Migrating project modules

  → Parsing Puppetfile at /home/lucy/githubs/bolt/project/Puppetfile

  → Select modules that are direct dependencies of your project. Bolt will
    automatically manage dependencies for each module selected, so do not select
    a module's dependencies unless you use content from it directly in your
    project.

    Select all modules? ([y]es/[n]o) y

  → Resolving module dependencies, this may take a moment

bundler: failed to load command: bolt (/home/lucy/githubs/bolt/vendor/bundle/ruby/2.6.0/bin/bolt)
ArgumentError: Module name (-canary) must match either 'modulename' or 'owner/modulename'
  /home/lucy/githubs/bolt/vendor/bundle/ruby/2.6.0/gems/puppetfile-resolver-0.4.0/lib/puppetfile-resolver/puppetfile/base_module.rb:57:in `parse_title'
  /home/lucy/githubs/bolt/vendor/bundle/ruby/2.6.0/gems/puppetfile-resolver-0.4.0/lib/puppetfile-resolver/puppetfile/base_module.rb:39:in `initialize'
  /home/lucy/githubs/bolt/vendor/bundle/ruby/2.6.0/gems/puppetfile-resolver-0.4.0/lib/puppetfile-resolver/puppetfile/forge_module.rb:9:in `initialize'
  /home/lucy/githubs/bolt/lib/bolt/puppetfile.rb:76:in `new'
  /home/lucy/githubs/bolt/lib/bolt/puppetfile.rb:76:in `block in resolve'
  /home/lucy/.rvm/rubies/ruby-2.6.3/lib/ruby/2.6.0/set.rb:338:in `each_key'
  /home/lucy/.rvm/rubies/ruby-2.6.3/lib/ruby/2.6.0/set.rb:338:in `each'
  /home/lucy/githubs/bolt/lib/bolt/puppetfile.rb:74:in `resolve'
  /home/lucy/githubs/bolt/lib/bolt/project_migrator/modules.rb:72:in `migrate_modules_from_puppetfile'
  /home/lucy/githubs/bolt/lib/bolt/project_migrator/modules.rb:27:in `migrate'
  /home/lucy/githubs/bolt/lib/bolt/project_migrator.rb:55:in `migrate_modules'
  /home/lucy/githubs/bolt/lib/bolt/project_migrator.rb:15:in `migrate'
  /home/lucy/githubs/bolt/lib/bolt/cli.rb:465:in `execute'
  /home/lucy/githubs/bolt/exe/bolt:11:in `<top (required)>'
  /home/lucy/githubs/bolt/vendor/bundle/ruby/2.6.0/bin/bolt:23:in `load'
  /home/lucy/githubs/bolt/vendor/bundle/ruby/2.6.0/bin/bolt:23:in `<top (required)>'
```
We should catch that and return just the error.

Reply:
Huh, that's weird... it should be catching all errors during the parse/resolve steps. It might be when it's building the document for resolving. Good find!

Reply:
Error stems from `Bolt::Puppetfile.parse` assuming all parsed modules are Forge modules. Updated the parser to error when a non-Forge module is present, unless the `skip_unsupported_modules` flag is set, in which case it will just ignore the unsupported modules. This is primarily to allow `project migrate` to still migrate a Puppetfile that has non-Forge modules.